src/services/factories/tokenList.ts

```diff
@@ -12,58 +12,106 @@ export function getTokensFactory(factoryParams: {
 
   const updatedIdsForNetwork = new Set<number>()
 
-  interface GetTokenIdResult {
-    value: number | null
+  interface UpdateTokenIdResult {
+    token: TokenDetails
     remove?: true
+    failed?: true
   }
 
-  async function getTokenId(networkId: number, tokenAddress: string): Promise<GetTokenIdResult> {
+  async function updateTokenId(networkId: number, token: TokenDetails): Promise<UpdateTokenIdResult> {
     try {
-      const hasToken = await exchangeApi.hasToken({ networkId, tokenAddress })
+      const hasToken = await exchangeApi.hasToken({ networkId, tokenAddress: token.address })
       if (!hasToken) {
-        return { value: null, remove: true }
+        return { token, remove: true }
       }
-      return { value: await exchangeApi.getTokenIdByAddress({ networkId, tokenAddress }) }
+      token.id = await exchangeApi.getTokenIdByAddress({ networkId, tokenAddress: token.address })
+      return { token }
     } catch (e) {
-      logDebug(`[network:${networkId}][address:${tokenAddress}] Failed to fetch id from contract`, e.message)
-      return { value: null }
+      logDebug(`[network:${networkId}][address:${token.address}] Failed to fetch id from contract`, e.message)
+      return { token, failed: true }
     }
   }
 
-  async function updateTokenIds(networkId: number): Promise<void> {
-    // Set as updated on start to prevent concurrent queries
-    updatedIdsForNetwork.add(networkId)
+  async function updateWithRetries(
+    networkId: number,
+    tokens: TokenDetails[],
+    maxRetries = 3,
+    retryDelay = 1000,
+    exponentialBackOff = true,
+  ): Promise<void> {
+    if (maxRetries <= 0) {
+      throw new Error(
+        `[network:${networkId}] Max retries exceeded trying to fetch token ids for tokens ${tokens.map(
+          t => t.address,
+        )}`,
```

Comment:
do we really want to hard-fail here, as opposed to a warning?
not that uncaught promise rejection is a very hard-fail, but still
we've already removed the always-failing token at this point anyway

Reply:
Not really hard failing. This exception is caught on https://github.com/gnosis/dex-react/pull/951/files#diff-e5e8192286e6ad1d544324c3dce8ed97R109

When it does fail, I clear the flag so next time the service is called we try to fetch again.

Reply:
Cool, good idea :+1: